### ISSUE TITLE
Selenium Node.js client passes string array to Element Send Keys, causing InvalidArgumentError: Could not convert 'text' to string

### ISSUE THREAD

**Issue by skratchdot** [mozilla/geckodriver#683]
In order to help us efficiently investigate your issue, please provide the following information:

## Firefox Version
53.0 (64-bit)

## Platform
macOS 10.12.4

## Additional Versions
geckodriver 16.0
selenium-webdriver 3.4.0

## Steps to reproduce -

I'm using the example script from: https://seleniumhq.github.io/selenium/docs/api/javascript/index.html with a slight modification to try to enable trace options.  I could not get the trace output to show no matter which format I tried (and I tried a bunch). Can anyone help on this front?

`brew cask install geckodriver`
`npm install selenium-webdriver@3.4.0`

`test-file.js`:
```
var webdriver = require("selenium-webdriver"),
  By = webdriver.By,
  until = webdriver.until;

var driver = new webdriver.Builder()
  .withCapabilities({
    browserName: "firefox",
    loggingPrefs: {
      driver: "ALL",
      performance: "ALL"
    },
    "moz:firefoxOptions": {
      prefs: {
        "dom.ipc.processCount": 8
      },
      log: {
        level: "trace"
      }
    }
  })
  .build();

driver.getCapabilities().then(console.log);

driver.get("http://www.google.com/ncr");
driver.findElement(By.name("q")).sendKeys("webdriver");
driver.findElement(By.name("btnG")).click();
driver.wait(until.titleIs("webdriver - Google Search"), 1000);
driver.quit();
```

```
$ node test-file.js
/Users/[MY_USERNAME]/[MY_PROJECT]/node_modules/selenium-webdriver/lib/promise.js:2634
        throw error;
        ^

InvalidArgumentError: Could not convert 'text' to string
    at Object.throwDecodedError (/Users/[MY_USERNAME]/[MY_PROJECT]/node_modules/selenium-webdriver/lib/error.js:489:11)
    at parseHttpResponse (/Users/[MY_USERNAME]/[MY_PROJECT]/node_modules/selenium-webdriver/lib/http.js:519:13)
    at doSend.then.response (/Users/[MY_USERNAME]/[MY_PROJECT]/node_modules/selenium-webdriver/lib/http.js:441:30)
    at process._tickCallback (internal/process/next_tick.js:109:7)
From: Task: WebElement.sendKeys()
    at thenableWebDriverProxy.schedule (/Users/[MY_USERNAME]/[MY_PROJECT]/node_modules/selenium-webdriver/lib/webdriver.js:816:17)
    at WebElementPromise.schedule_ (/Users/[MY_USERNAME]/[MY_PROJECT]/node_modules/selenium-webdriver/lib/webdriver.js:2024:25)
    at WebElementPromise.sendKeys (/Users/[MY_USERNAME]/[MY_PROJECT]/node_modules/selenium-webdriver/lib/webdriver.js:2187:19)
    at Object.<anonymous> (/Users/[MY_USERNAME]/[MY_PROJECT]/test-file.js:10:34)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.runMain (module.js:605:10)
```

 - [x] A *minimal* HTML file (or files) that can be used to reproduce the issue (this should not be your production site, unless you are unable to isolate the problem, but a page containing only the elements required to observe the problem).

 - [x] A [trace level log](https://github.com/mozilla/geckodriver#firefox-capabilities) from a geckodriver session that demonstrates the problem using the minimal example file. Note: this is *not* the same as a selenium log. (see note: I could not get this working).

**Comment by skratchdot** [comment]
I'm only logging this here b/c the selenium repo points here.  It could very well just be a bug w/ [selenium-webdriver](https://github.com/SeleniumHQ/selenium/issues) but I figured I'd at least log the issue here so we could cross-reference things

**Comment by andreastt** [comment]
It’s hard to tell if this is a bug in Selenium unless you attach the trace-level geckodriver log. From the error message, I suspect your language binding is sending the wrong data type to Element Send Keys, by passing a `{text: ["f", "o", "o"]}` (array) object, whereas geckodriver is expecting `{text: "foo"}` (string).

**Comment by skratchdot** [comment]
Thanks @andreastt!  Do you have any experience with `selenium-webdriver` in node?  I tried everything I could think of to pass ```log: {
        level: "trace"
      }``` to geckodriver, but I just couldn't figure it out.  I'm glad you didn't automatically close this b/c I wanted to attach the trace-level file, but couldn't figure it out.  I'll try to look at this again after work and see if I can get a trace file (or actually debug the `node_modules/selenium-webdriver/lib/webdriver.js:2187:19` line).

**Comment by andreastt** [comment]
Paging @jleyba: Can you tell us how to pass {"moz:firefoxOptions": {"log": {"level": "trace"}}}` through the Node.js client? I think I will create a wiki page in geckodriver on how to do this.

**Comment by skratchdot** [comment]
@andreastt - I figured it out.  I tried a different method. In one tab I ran `geckodriver --log trace`, then I changed my builder code from:

```
var driver = new webdriver.Builder()
  .withCapabilities({ ... })
  .build()
```
to:
```
var driver = new webdriver.Builder()
  .forBrowser('firefox')
  .usingServer('http://localhost:4444/')
  .build();
```

Anyways, here is the trace file:

```
1493162411377	webdriver::httpapi	DEBUG	Creating routes
1493162411419	geckodriver	INFO	Listening on 127.0.0.1:4444
1493162413976	webdriver::server	DEBUG	→ POST /session {"desiredCapabilities":{"browserName":"firefox","moz:firefoxOptions":{}}}
1493162414020	geckodriver::marionette	INFO	Starting browser /Applications/Firefox.app/Contents/MacOS/firefox-bin with args ["-marionette"]
1493162414024	geckodriver::marionette	TRACE	  connection attempt 0/600
1493162414127	geckodriver::marionette	TRACE	  connection attempt 1/600
1493162414229	geckodriver::marionette	TRACE	  connection attempt 2/600
1493162414334	geckodriver::marionette	TRACE	  connection attempt 3/600
1493162414436	geckodriver::marionette	TRACE	  connection attempt 4/600
1493162414537	Marionette	DEBUG	Marionette enabled via command-line flag
1493162414538	geckodriver::marionette	TRACE	  connection attempt 5/600
1493162414643	geckodriver::marionette	TRACE	  connection attempt 6/600
1493162414748	geckodriver::marionette	TRACE	  connection attempt 7/600
1493162414853	geckodriver::marionette	TRACE	  connection attempt 8/600
1493162414956	geckodriver::marionette	TRACE	  connection attempt 9/600
1493162415061	geckodriver::marionette	TRACE	  connection attempt 10/600
1493162415163	geckodriver::marionette	TRACE	  connection attempt 11/600
1493162415268	geckodriver::marionette	TRACE	  connection attempt 12/600
1493162415371	geckodriver::marionette	TRACE	  connection attempt 13/600
1493162415429	Marionette	INFO	Listening on port 51159
1493162415473	geckodriver::marionette	DEBUG	Connected to Marionette on localhost:51159
1493162415488	Marionette	DEBUG	Accepted connection conn0 from 127.0.0.1:51194
1493162415492	geckodriver::marionette	TRACE	← {"applicationType":"gecko","marionetteProtocol":3}
1493162415492	geckodriver::marionette	TRACE	→ 109:[0,1,"newSession",{"browserName":"firefox","capabilities":{"desiredCapabilities":{"browserName":"firefox"}}}]
1493162415495	Marionette	TRACE	conn0 -> [0,1,"newSession",{"browserName":"firefox","capabilities":{"desiredCapabilities":{"browserName":"firefox"}}}]
1493162415498	Marionette	CONFIG	Matched capabilities: {"browserName":"firefox","browserVersion":"53.0","platformName":"darwin","platformVersion":"16.5.0","pageLoadStrategy":"normal","acceptInsecureCerts":false,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000},"rotatable":false,"specificationLevel":0,"moz:processID":80496,"moz:profile":"/var/folders/d2/rvs5kqdj2fqg1k9qxp0vdp3r3kfx1h/T/rust_mozprofile.sJYO3wZ1AJBg","moz:accessibilityChecks":false}
1493162415731	Marionette	TRACE	conn0 <- [1,1,null,{"sessionId":"ee4b7795-17d1-4f40-8146-1e469744025d","capabilities":{"browserName":"firefox","browserVersion":"53.0","platformName":"darwin","platformVersion":"16.5.0","pageLoadStrategy":"normal","acceptInsecureCerts":false,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000},"rotatable":false,"specificationLevel":0,"moz:processID":80496,"moz:profile":"/var/folders/d2/rvs5kqdj2fqg1k9qxp0vdp3r3kfx1h/T/rust_mozprofile.sJYO3wZ1AJBg","moz:accessibilityChecks":false}}]
1493162415732	geckodriver::marionette	TRACE	← [1,1,null,{"sessionId":"ee4b7795-17d1-4f40-8146-1e469744025d","capabilities":{"browserName":"firefox","browserVersion":"53.0","platformName":"darwin","platformVersion":"16.5.0","pageLoadStrategy":"normal","acceptInsecureCerts":false,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000},"rotatable":false,"specificationLevel":0,"moz:processID":80496,"moz:profile":"/var/folders/d2/rvs5kqdj2fqg1k9qxp0vdp3r3kfx1h/T/rust_mozprofile.sJYO3wZ1AJBg","moz:accessibilityChecks":false}}]
1493162415733	webdriver::server	DEBUG	← 200 OK {"value": {"sessionId":"ee4b7795-17d1-4f40-8146-1e469744025d","capabilities":{"acceptInsecureCerts":false,"browserName":"firefox","browserVersion":"53.0","moz:accessibilityChecks":false,"moz:processID":80496,"moz:profile":"/var/folders/d2/rvs5kqdj2fqg1k9qxp0vdp3r3kfx1h/T/rust_mozprofile.sJYO3wZ1AJBg","pageLoadStrategy":"normal","platformName":"darwin","platformVersion":"16.5.0","rotatable":false,"specificationLevel":0,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000}}}}
1493162415744	webdriver::server	DEBUG	→ POST /session/ee4b7795-17d1-4f40-8146-1e469744025d/url {"url":"http://www.google.com/ncr"}
1493162415745	geckodriver::marionette	TRACE	→ 47:[0,2,"get",{"url":"http://www.google.com/ncr"}]
1493162415746	Marionette	TRACE	conn0 -> [0,2,"get",{"url":"http://www.google.com/ncr"}]
2017-04-25 19:20:15.791 plugin-container[80497:1319475] *** CFMessagePort: bootstrap_register(): failed 1100 (0x44c) 'Permission denied', port = 0x9f3b, name = 'com.apple.tsm.portname'
See /usr/include/servers/bootstrap_defs.h for the error codes.
2017-04-25 19:20:15.792 plugin-container[80497:1319475] *** CFMessagePort: bootstrap_register(): failed 1100 (0x44c) 'Permission denied', port = 0xa203, name = 'com.apple.CFPasteboardClient'
See /usr/include/servers/bootstrap_defs.h for the error codes.
1493162417841	Marionette	TRACE	conn0 <- [1,2,null,{}]
1493162417843	geckodriver::marionette	TRACE	← [1,2,null,{}]
1493162417843	webdriver::server	DEBUG	← 200 OK {"value": {}}
1493162417846	webdriver::server	DEBUG	→ POST /session/ee4b7795-17d1-4f40-8146-1e469744025d/element {"using":"css selector","value":"*[name=\"q\"]"}
1493162417849	geckodriver::marionette	TRACE	→ 68:[0,3,"findElement",{"using":"css selector","value":"*[name=\"q\"]"}]
1493162417850	Marionette	TRACE	conn0 -> [0,3,"findElement",{"using":"css selector","value":"*[name=\"q\"]"}]
1493162417854	Marionette	TRACE	conn0 <- [1,3,null,{"value":{"element-6066-11e4-a52e-4f735466cecf":"e057385b-b4f4-c444-b919-dcd6011c7803","ELEMENT":"e057385b-b4f4-c444-b919-dcd6011c7803"}}]
1493162417855	geckodriver::marionette	TRACE	← [1,3,null,{"value":{"element-6066-11e4-a52e-4f735466cecf":"e057385b-b4f4-c444-b919-dcd6011c7803","ELEMENT":"e057385b-b4f4-c444-b919-dcd6011c7803"}}]
1493162417855	webdriver::server	DEBUG	← 200 OK {"value":{"element-6066-11e4-a52e-4f735466cecf":"e057385b-b4f4-c444-b919-dcd6011c7803"}}
1493162417858	webdriver::server	DEBUG	→ POST /session/ee4b7795-17d1-4f40-8146-1e469744025d/element/e057385b-b4f4-c444-b919-dcd6011c7803/value {"text":["w","e","b","d","r","i","v","e","r"],"value":["w","e","b","d","r","i","v","e","r"]}
1493162417872	webdriver::server	DEBUG	← 400 Bad Request {"value":{"error":"invalid argument","message":"Could not convert 'text' to string","stacktrace":"stack backtrace:\n   0:        0x104e51ea1 - backtrace::backtrace::trace::h5db1675e0d2383fc\n   1:        0x104e53394 - backtrace::capture::Backtrace::new::h8ca3ad60a3bf61a1\n   2:        0x104e44d61 - webdriver::error::WebDriverError::new::hb54bb5c3098fbb06\n   3:        0x104e370f5 - _$LT$webdriver..command..SendKeysParameters$u20$as$u20$webdriver..command..Parameters$GT$::from_json::hceb2bdbf2b6bdfc6\n   4:        0x104d66670 - _$LT$webdriver..command..WebDriverMessage$LT$U$GT$$GT$::from_http::hadee1455d862898a\n   5:        0x104d6921f - _$LT$webdriver..httpapi..WebDriverHttpApi$LT$U$GT$$GT$::decode_request::haf6f925c1be8a3fd\n   6:        0x104dab3c9 - _$LT$webdriver..server..HttpHandler$LT$U$GT$$u20$as$u20$hyper..server..Handler$GT$::handle::hc13164f59a2c525a\n   7:        0x104ce0dd9 - _$LT$hyper..server..Worker$LT$H$GT$$GT$::keep_alive_loop::hd9e10eb5bdab56da\n   8:        0x104ce1a23 - _$LT$hyper..server..Worker$LT$H$GT$$GT$::handle_connection::h8e1cc6272996cc0f\n   9:        0x104d76817 - hyper::server::handle::_$u7b$$u7b$closure$u7d$$u7d$::h44ee354f6bc152f7\n  10:        0x104d76d77 - hyper::server::listener::spawn_with::_$u7b$$u7b$closure$u7d$$u7d$::hb1c5c489635a6da2\n  11:        0x104dba58e - _$LT$std..panic..AssertUnwindSafe$LT$F$GT$$u20$as$u20$core..ops..FnOnce$LT$$LP$$RP$$GT$$GT$::call_once::h1f988be5bac5120c\n  12:        0x104cf52ad - std::panicking::try::do_call::h75b3846b2d36424a\n  13:        0x1053ade8a - __rust_maybe_catch_panic\n  14:        0x104cf4580 - std::panicking::try::h1a4234eb1dae2f13\n  15:        0x104cf17c8 - std::panic::catch_unwind::h7519976f6c64a0f5\n  16:        0x104cf3ec2 - std::thread::Builder::spawn::_$u7b$$u7b$closure$u7d$$u7d$::h5e63abb2a16dcb23\n  17:        0x104d51626 - _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::ha941dc3eec0e8667\n  18:        0x1053ac2b4 - std::sys::imp::thread::Thread::new::thread_start::hf2762ec0b24c4077\n  19:     0x7fffa8d729ae - _pthread_body\n  20:     0x7fffa8d728fa - _pthread_start"}}
1493162423773	Marionette	DEBUG	Closed connection conn0
```

**Comment by skratchdot** [comment]
So you are correct about the `text` being an array.  I guess this is a bug w/ selenium-webdriver.

**Comment by skratchdot** [comment]
There is a comment here: https://github.com/SeleniumHQ/selenium/blame/a709f452b12413123e865d99e2ce936569d818ff/javascript/node/selenium-webdriver/lib/webdriver.js#L2179 stating `The W3C protocol requires keys to be specified as an array where each element is a single key.`

I'm guessing since this is a year old, it is no longer the case w/ the w3c spec?

I'm trying to find out by reading: https://w3c.github.io/webdriver/webdriver-spec.html#element-send-keys

**Comment by andreastt** [comment]
That’s correct, the spec recently changed to accept a string instead of an array-string for Element Send Keys, so if that comment is over a year old, this is a case of fixing the Node.js client binding to pass a string.

**Comment by andreastt** [comment]
@skratchdot Say, speaking of which, you seem pretty proficient debugging this stuff, do you want to have a go at fixing the Selenium Node.js codec or at least filing a bug there?

**Comment by skratchdot** [comment]
Sure.  I just made a fix that seems to work in both the chromedriver and geckodriver, but I want to do a bit more testing, and I need to try to figure out how to make sure I'm not breaking any of the `sendKeys()` functionality for "special characters", etc.  I'll post an update back here when the upstream node PR is submitted.  Thanks for your help!

**Comment by skratchdot** [comment]
I'm going to close this since the "fix" I put in to the upstream node.js lib was merged in.  Fingers crossed that I don't break a bunch of stuff for people.  If you want to re-open for better visibility, feel free!

**Comment by andreastt** [comment]
@skratchdot Thanks for patching the bindings!

**Comment by ShellOliver** [comment]
The same thing happens yet, but using geckodriver 16.1 on linux.
I've tested using opera driver and it works, I also used the example in https://seleniumhq.github.io/selenium/docs/api/javascript/:
`var webdriver = require('selenium-webdriver'),
    By = webdriver.By,
    until = webdriver.until;

var driver = new webdriver.Builder()
    .forBrowser('firefox')
    .build();

driver.get('http://www.google.com/ncr');
driver.findElement(By.name('q')).sendKeys('webdriver');
driver.findElement(By.name('btnG')).click();
driver.wait(until.titleIs('webdriver - Google Search'), 1000);
driver.quit();`

**Comment by rlora** [comment]
@ShellOliver I think the merge was made to the `webdriver`. Currently it is on version `3.4.0` dated April 21. I'm also experiencing the issue and was waiting for `3.4.1` to be released. Although I'm not 100% percent sure that is the place with the fix. Maybe @skratchdot can confirm? Thanks!!

**Comment by skratchdot** [comment]
@rlora / @ShellOliver - that's correct.  They merged the fix in, but have not yet published a new version.  You can test by running the following command in your project root:

```
sed '2189q;d' ./node_modules/selenium-webdriver/lib/webdriver.js
```

The old line 2189 looks like:
```
setParameter('text', keys).
```
The new line 2189 will look like:
```
setParameter('text', keys.then(keys => keys.join(''))).
```

You can try editing your local install to see if things start working for you.  I really hope this doesn't break stuff for people.  I *think* it should be okay, but I didn't do extensive testing (and am not sure of all the `sendKeys()` usage types / edge cases).

So if you get a chance, please try making the above edit, and re-running your tests!

**Comment by rlora** [comment]
Hi @skratchdot, writing to confirm that the change works for me. Thank you!

```
MacOs Sierra 10.12.4
node 6.9.5, selenium-webdriver 3.4.0 + patch
geckodriver 0.16.1
Firefox 53
```

**Comment by skratchdot** [comment]
Nice! Thanks for taking the time to test!

**Comment by ShellOliver** [comment]
Thanks @skratchdot and @rlora ! it really works, tested using firefox and chrome.

**Comment by obitwo** [comment]
@skratchdot Exactly what I needed. Thanks.

**Comment by santutu** [comment]
@skratchdot oh.. thanks , this is i needed...  i resolve easliy problem

**Comment by sschuberth** [comment]
So, how can I work around the issue until a new version of selenium-webdriver is released?

**Comment by skratchdot** [comment]
@sschuberth - You could probably have a postinstall script that applies a patch.  Or you could just ask the selenium team when their next release will be (or ask if they will consider a making minor node release prior to their scheduled release).  I was going to suggest just installing from `SeleniumHQ/selenium#master`, but I don't think that will work since they are using a monorepo, and there is no `package.json` in the root of that project at github.  There might be a workaround, but after looking at https://docs.npmjs.com/files/package.json#git-urls-as-dependencies it seems like you could use "local paths" instead of a github url.  I don't know if that will work for your usecase though.

**Comment by sschuberth** [comment]
>  I was going to suggest just installing from SeleniumHQ/selenium#master, but I don't think that will work since they are using a monorepo, and there is no package.json in the root of that project at github.

Yeah, I already tried that, and figured that out the hard way 😞 

I think I'll go for the patch file then. Thanks!

**Comment by kandaj** [comment]
can someone help me on how to apply patch in javascript

**Comment by shaglou** [comment]
patched selenium-webdriver 3.4.0 with this specific change, anyone is welcome to use the patched version here via: 
npm i selenium-webdriver-patched --save-dev
LOL

**Comment by VedemV** [comment]
Hi @skratchdot, writing to confirm that the change works for me. Thank you!
`Windows 10,
node 6.9.1, 
selenium-webdriver 3.4.0,
geckodriver 0.17.0
Firefox 54`
